rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+<code>example_gen = CsvExampleGen(input_base=examples)</code>
+<p>
+<code><em># Note that the output will be of type 'ExamplesPath', for which we</em></code>
+<code><em># may have registered a notebook visualization handler.</em></code>
+<code>example_gen_result = context.run(example_gen)</code>
+<p>
+<code>example_gen.outputs['examples']</code>
+<p>
+<code><em># alternative style: explicit context.show() method</em></code>
+<code>context.show(example_gen.outputs['examples'])</code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[1]:</strong>
+<p>
+_notebook visualization indicating we have N examples at some temp path_
+   </td>
+  </tr>
+</table>
+
+<table>
+  <tr>
+   <td><strong>Input[2]:</strong>
+<p>
+<code><em># Next, StatisticsGen with a run / read.</em></code>
+<code>statistics_gen = StatisticsGen(input_data=example_gen.outputs['examples'])</code>
+<p>
+<code>context.run(statistics_gen).outputs['output']</code>
+<p>
+<code><em># alternative styles:</em></code>
+<code><em># context.<strong>show(</strong>context.run(statistics_gen).outputs['output'])</em></code>
+<code><em># context.run().read('output'<strong>, visualization_handler=blah</strong>)</em></code>
+<code><em># context.run().<strong>show</strong>('output', <strong>visualization_handler=blah, visualization_args=</strong>)</em></code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[2]:</strong>
+<p>
+_notebook visualization for statsgen output_
+   </td>
+  </tr>
+</table>
+
+<table>
+  <tr>
+   <td><strong>Input[3]:</strong>
+<p>
+<code><em># Next, SchemaGen without a run / read.</em></code>
+<code>infer_schema = SchemaGen(stats=statistics_gen.outputs['output'])</code>
+<p>
+<code><em># Finally, ExampleValidator with a run / read. Note that SchemaGen</em></code>
+<code><em># will be implicitly run (see Note 2 below).</em></code>
+<code>validate_stats = ExampleValidator(</code>
+<p>
+<code>    stats=statistics_gen.outputs['output'],</code>
+<p>
+<code>    schema=infer_schema.outputs['output'])</code>
+<p>
+<code>context.run(validate_stats)</code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[3]:</strong>
+<p>
+_ExecutionResult object for ExampleValidator_
+   </td>
+  </tr>
+</table>
+
+Note that the user may have forgotten to run InteractiveContext.run() on
+upstream components in the dependency graph. Instead of implicitly running these
+upstream components, we remind the user to explicitly run upstream notebook
+cells (with a readable error message). We think this explicit mode of component
+execution is more notebook-friendly, and is easy to use with common notebook
+actions such as “Run All”, “Run Cells Before”, and “Run Cells After”.
+
+### Export to a selected orchestration engine (v0)
+
+#### Filter out InteractiveContext() objects
+
+##### Option 1: Replace InteractiveContext instances with dummy versions.
+
+1.  Search for possible import alias, e.g. `from
+    tfx.orchestration.interactive.interactive_context import InteractiveContext
```

Comment:
Who is doing this search? Can you clarify the context of the process being described here? Does this imply that there would be some process by which we do this translation for users? Seems fragile. We do a lot of Python import magic to make TF exports work, and I wouldn't recommend doing that in arbitrary user code.